pandas/core/algorithms.py

```diff
@@ -989,9 +989,10 @@ def _get_data_algo(values, func_map):
         f = func_map['uint64']
         values = _ensure_uint64(values)
 
-    elif is_categorical_dtype(values):
-        f = func_map['int64']
-        values = _ensure_int64(values.codes)
+    elif is_categorical_dtype(values) and values.ordered:
```

Comment:
as I said I'd like this to be actually in ``Categorical``. and don't use ``np.vectorize``. use masking.

something like (inside Categorical)

```
codes = self.codes
mask = codes == -1
if mask.any():
    codes = codes.astype('float64')
    codes[mask] = np.nan
```